Focused Launch: Implement step highlighting logic

alshakero opened the pull request:
#### Changes proposed in this Pull Request

* Implements highlighting logic for each step in the Summary view.

#### Testing instructions

**Important**: Please remember to reset localStorage when trying to clear some pre-selected values when opening and re-openeing the Focus Launch modal.


1. Create a new site via `/start` giving it no title and ensuring it's on the free plan.
2. Go to `/page/[UNLAUNCHED_SITE_CREATED_BY_START]/home?flags=create/focused-launch-flow-calypso`
3. Click "Launch" to open the modal


- [x] Both the Domains Step and the Plans Step should be disabled at this point
- [x] Enter a site title, making sure it's not "Site Title", this should enable the Domains Step
- [x] Change the site title back to "Site Title" to check this disables the Domains Step again
- [x] Making sure the Domains Step is enabled (repeat step 5) no domain (not even the sub-domain) should be pre-selected. The Plans Step should _still_ be disabled at this point.
- [x] Selecting a domain should highlight it as selected, and should now enable the plans step.
- [x] Selecting a plan should highlight it as selected. At this point all steps should be enabled.

Fixes #47458
Fixes #47743

alshakero committed:
Implement step highlighting logic

Commit:
Add conditions to display highlighted state for each step.

Commit:
Display site name step if title is empty string

wp-desktop reviewed (dismissed):
WordPress Desktop CI Failure for job "wp-desktop-mac".

@razvanpapadopol please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/959096). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

tjcafferkey commented:
I have tested this on the `.live` domain as well as checking out the branch locally and this is what happened to me. Please let me know if I am doing anything wrong RE testing steps.

#### If I already had a site title & subdomain

If I already had a site title and subdomain, it pre-selected my subdomain (which is what I would expect) and showed the plans step as inactive. I was happy with my subdomain choice already so I clicked it again to enable the plans step. Nothing happened.

[Video here](https://www.loom.com/share/6f85262f6bde4b4daa22cd562c885cbc)

#### If I had no site title it enabled my domain step

If I setup a new site with no title and skipped every possible step and then landed in the focused launch my title field was pre-filled and my domain step was enabled with my sub-domain selected. Is this expected?

[Video here](https://www.loom.com/share/5a3927d6b8e94acd9dc91b7500a30039)

#### Gutenboarding setup and focus launch enabled all steps

This seemed inconsistent as I expected it to be like the above but I video'd the experience from site creation and my site name was prefilled with "Site Title" and all steps were enabled

[Video here](https://www.loom.com/share/2fd4b391a48c4b32b11feef788d3801c)

razvanpapadopol commented:
> **If I already had a site title & subdomain**
>
> If I already had a site title it pre-selected my subdomain (which is what I would expect) and showed the domain step as inactive. I was happy with my subdomain choice already so I clicked it again to enable the domain step. Nothing happened.

The subdomain shouldn't be pre-selected as specified in https://github.com/Automattic/wp-calypso/issues/47540 (last acceptance criteria).
So I expect this issue to be solved in https://github.com/Automattic/wp-calypso/pull/47633 or in a follow up.

> **If I had no site title it enabled my domain step**
>
> If I setup a new site with no title and skipped every possible step and then landed in the focused launch my title field was pre-filled and my domain step was enabled with my sub-domain selected. Is this expected?

Yes, the site title step has a value even if it's the default value. Here is the design for it:

<img width="607" alt="Screenshot 2020-11-25 at 12 22 34" src="https://user-images.githubusercontent.com/14192054/100214874-0c2e9780-2f19-11eb-877a-a35a66d65a40.png">

> **Gutenboarding setup and focus launch enabled all steps**
>
> Gutenboarding setup and focus launch enabled all steps
This seemed inconsistent as I expected it to be like the above but I video'd the experience from site creation and my site name was prefilled with "Site Title" and all steps were enabled

You probably didn't clear local storage data. When using the Calypso modal (for testing only at this point), Launch data-store is persisted between different sites since the domain is the same. Issue: https://github.com/Automattic/wp-calypso/issues/47328

*Note: you don't need to sandbox when testing changes exclusively with `?flags=create/focused-launch-flow-calypso` flag since that modal is 100% in Calypso (no Editing Toolkit plugin code is executed).

tjcafferkey commented:
>The subdomain shouldn't be pre-selected as specified in #47540 (last acceptance criteria). So I expect this issue to be solved in #47633 or in a follow up.

I think we probably shouldn't cross post acceptance criteria (A/C) on issues as this can cause dependencies between issues (e.g. X pr will meet it's acceptance criteria once Y pr has been merged, as it's being worked on there). Instead it feels more relevant to list it as a 'consideration' (for lack of a better term) on one PR and in the acceptance criteria of the most appropriate one where it can be referenced. 

Because of this, it's been removed from the A/C of https://github.com/Automattic/wp-calypso/issues/47540 to under the consideration heading. This is probably more appropriate as that particular A/C only applies in context of the focused launch where as issue https://github.com/Automattic/wp-calypso/issues/47540 is tackling domain picker functionality irrespective of the context it's used.

Would you agree @razvanpapadopol ?

razvanpapadopol commented:
> Because of this, it's been removed from the A/C of #47540 to under the consideration heading. This is probably more appropriate as that particular A/C only applies in context of the focused launch where as issue #47540 is tackling domain picker functionality irrespective of the context it's used.
>
> Would you agree @razvanpapadopol ?

Yes, 100%. Probably it was included there as a known domain-related issue that needed to be fixed. It's also not related to highlighting steps so I wouldn't include it in this PR. Could you create a dedicated issue for it (eg: "Focused Launch: don't pre-select subdomain")?

tjcafferkey committed:
Merge branch 'trunk' into update/highlight-steps

tjcafferkey committed:
Remove duplicate selector

tjcafferkey committed:
Remove unused React imports from Summary view

tjcafferkey committed:
Remove sitePrimaryDomain as fallback currentDomain for DomainPicker in Summary view. This prevents the default selection being sitePrimaryDomain

tjcafferkey committed:
Revert "Remove sitePrimaryDomain as fallback currentDomain for DomainPicker in Summary view. This prevents the default selection being sitePrimaryDomain"

This reverts commit 578ac32cc22f8f8aec8b811a78d3795489ce1765.

Commit:
Focused launch: Highlight flow steps [part 2] (#47912)

Co-authored-by: Tom Cafferkey <tjcafferkey@gmail.com>

tjcafferkey committed:
Change isValidSiteTitle to check the title isn't exactly like the default title.

This is because any title with "Site Title" in it will be invalid. We should be satisfied with anything the user changes it to that isn't the default title. Even if its a variation of "Site Title"

razvanpapadopol commented on packages/launch/src/utils.ts:
```diff
@@ -22,7 +22,7 @@ export const isDefaultSiteTitle = ( {
 
 export const isValidSiteTitle = ( title?: string ): boolean =>
-	title !== '' && ! isDefaultSiteTitle( { currentSiteTitle: title } );
+	title !== '' && ! isDefaultSiteTitle( { currentSiteTitle: title, exact: true } );
```
This change looks right. Since now we're calling everywhere `isDefaultSiteTitle` with the `exact: true` option, maybe that's not needed anymore 🤷

tjcafferkey replied:
Probably right but it could be nice to have that optional flexibility on this function? I can remove it if we feel it's completely redundant now. What do you think?

razvanpapadopol replied:
I think this has been added to prevent the layout from jumping when we didn't have the site title step persistent. 

There is one very specific case I have in mind: if a user adds some character to the default title, then closes the flow and opens it on a new device/browser (or clears local storage data). Only in this case when we might want to do the non-exact matching to show them the title step. But the scenario is very unlikely so we could remove this without any concern.

razvanpapadopol commented on packages/launch/src/hooks/use-domain-search.ts:
```diff
@@ -17,13 +18,15 @@ export function useDomainSearch(): {
+	const { isLoadingSite } = useSite();
+	const { siteSubdomain } = useSiteDomains();
+
 	const { setDomainSearch } = useDispatch( LAUNCH_STORE );
 
 	let search = domainSearch.trim() || title;
 
 	if ( ! search || isDefaultSiteTitle( { currentSiteTitle: search, exact: true } ) ) {
-		search = currentDomainName?.split( '.' )[ 0 ] ?? '';
+		search = siteSubdomain?.domain?.split( '.' )[ 0 ] ?? '';
```
This logic might need a bit of love to become more readable. The idea is use one of the following for domain search, in this order:
1. `domainSearch`
2. `title` if valid
3. `siteSubdomain?.domain?.split( '.' )[ 0 ]`
4. `''`

Also, it feels like `isValidSiteTitle` would be a better choice than `isDefaultSiteTitle`.

However, this isn't blocking for this PR.

tjcafferkey replied:
I will create a janitorial type issue for this.

tjcafferkey committed:
Move highlighted prop to CommonStepProps and move the logic to the SummaryStep.

The values needed to know which step is highlighted were already being used in the Summary component, and this was a common property shared amongst all steps. It didnt make sense to query those same values again in each individual component but rather let the Summary component control this logic.

Commit:
remove comment

Commit:
remove redundant initialization with undefined

Commit:
Merge branch 'trunk' into update/highlight-steps

tjcafferkey committed:
Remove exact argument from isDefaultSiteTitle utility function

ciampo reviewed (commented):
The changes test well as per instructions!

Although I just found another edge case:

1. Open Focused Launch modal
2. Change Site title — domain step is highlighted
3. Pick a domain — plan step is highlighted
4. Change Site title back to "Site Title"
5. At this point, the UI goes a bit _out of sync_:
   - The domain step is visually disabled again, but while a domain is selected
   - The plan step is enabled

![Screenshot 2020-12-03 at 11 34 40](https://user-images.githubusercontent.com/1083581/101000750-9781db80-355e-11eb-89a0-2a8bd8383124.png)


We should probably:
- change the `highlighted` prop to `false` for a step only if that step doesn't have a selected item yet
   - this would prevent the domain step from fading in the scenario above
- change the `highlighted` prop to `true` for a step, only if all previous steps are also highlighted
   - this would prevent the plan step from being highlighted in the scenario above

ciampo commented on packages/launch/src/utils.ts:
```diff
@@ -12,17 +12,12 @@ const DEFAULT_SITE_NAME = __( 'Site Title', __i18n_text_domain__ );
 // to change it before launch.
 export const isDefaultSiteTitle = ( {
 	currentSiteTitle = '',
-	exact = false,
```
We should probably update the comment above this function too

ciampo commented on apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-steps/domain-step/index.tsx:
```diff
@@ -14,20 +14,20 @@ import { recordTracksEvent } from '@automattic/calypso-analytics';
 	const { setDomain, unsetDomain, unsetPlan, confirmDomainSelection } = useDispatch( LAUNCH_STORE );
 
 	const handleNext = () => {
-		confirmDomainSelection();
```
Hey @razvanpapadopol , would you mind giving some context around these changes? They seem to affect Step by Step flow, but in the PR description I couldn't find any mention of them, and so I don't know what to test for

razvanpapadopol replied:
It's the same change for both flows: 
> Don't preselect subdomain anymore. An explicit selection is now needed in order for Plans step to be highlighted 

and it was reviewed and merged in https://github.com/Automattic/wp-calypso/pull/47912

In Step by Step flow, before this change, the user was able to confirm the domain selection by clicking Next button. Now since the subdomain is not pre-selected anymore, they need to select it.

ciampo replied:
Got it, thank you!

tjcafferkey committed:
Remove comment which refers to a function argument that no longer exists

ciampo commented on packages/launch/src/focused-launch/summary/style.scss:
```diff
@@ -82,20 +82,12 @@ $commentary-column-gap-wide: 100px;
-// by default: highlight all steps when nothing is in focus
-.focused-launch-container:not( :focus-within ) {
-	.focused-launch-summary__step {
+	&.highlighted {
```
Probably a shorter `transition-duration` of `0.2s` would make the animation feel a bit snappier? (see line 76 in this same file)

tjcafferkey committed:
Check the previous step has been completed, or the current step has a selected item in order to enable highlighting

tjcafferkey committed:
Speed up highlighting opacity

razvanpapadopol commented on packages/launch/src/focused-launch/summary/index.tsx:
```diff
@@ -598,9 +604,11 @@ const Summary: React.FunctionComponent = () => {
 		/>
 	);
 
+	const isPlansStepHighlighted = !! hasSelectedDomain || !! selectedPlan;
```
```suggestion
	const isPlansStepHighlighted = isValidSiteTitle( title ) && !! hasSelectedDomain;
```

I think the plan selection is not important at this step and we just need something like this (including title validation) in order to prevent this case:
<img width="562" alt="Screenshot 2020-12-03 at 16 16 17" src="https://user-images.githubusercontent.com/14192054/101034653-37516080-3583-11eb-835a-f0b3846d02f8.png">

tjcafferkey replied:
Me and @ciampo have just gone through these two scenarios. It's an edge case as it stands but we agreed that the current step should only be highlighted if the previous step has been completed, and not base it on _all_ previous steps.

This means that the user can only highlight the plans step if they select a domain (which they can only select if they've entered a valid title). If the user then decides to go back and change their title to the default title it doesn't seem right to then to 'disable' all the steps. 

Do you agree @razvanpapadopol or do you think we should only highlight the domains step if _all_ previous steps are completed and valid?

razvanpapadopol replied:
In this case user had a default Site title (like all users coming from `/start`) and Premium plan pre-selected (coming from cart; see https://github.com/Automattic/wp-calypso/pull/47990)

tjcafferkey replied:
@razvanpapadopol I've implemented your suggested changes. Worth noting now that in this case a user had a default Site title and domain pre-selected (coming from cart) then the plans step will not be highlighted.

razvanpapadopol replied:
🤦 then leave the highlight condition for Plans step [as it was](https://github.com/Automattic/wp-calypso/pull/47503/commits/856bb448254629add939d9454bc9dec3b508e168#diff-bae2915b59f2e567e179b1ed48bdcb4d210fbc99442732fa6252d32ef7c2c181L603), taking into consideration just the completion status of previous step and completely ignore title?

ciampo replied:
Having a domain from cart but not a plan is very unlikely scenario, so it's probably the lesser evil in this situation.

We should still add these notes to a separate issue, to discuss a final approach that aims at covering all edge cases

tjcafferkey replied:
Then there's the other scenario that if they have both a plan and a domain pre-selected from cart, the plans step will be disabled because the title is still the default Site Title.

tjcafferkey replied:
I have reverted this change so the user must have a selected domain or a selected plan for the plans step to be highlighted.

tjcafferkey committed:
Merge branch 'trunk' into update/highlight-steps

tjcafferkey committed:
Fix errors from merge conflict

tjcafferkey committed:
User has to have selected domain and a valid title for plans step to be enabled

tjcafferkey committed:
User must have a selected domain, or a selected plan for the plans step to be highlighted

razvanpapadopol reviewed (approved):
Ok, everything seems to be working as described in the spec and the UI feels responsive 👌 
Let's 🚢 this and add a follow-up issue with a design question as mentioned in https://github.com/Automattic/wp-calypso/pull/47503#discussion_r535335632
